articles/types/jsdoc.md

```diff
@@ -1,3 +1,179 @@
+
+<img src="../../images/jsdoc-quickinfo.png" width="1000px"/>
+
+and while filling in the paramters in a function call.
+
+<img src="../../images/jsdoc-param.png" width="1000px"/>
+
+Different constructs in JavaScript can use different forms of documentation, as described below.
+
+Visual Studio offers a rich editing experience inside JsDoc comments. The editor offers auto-completion on tags, names, and types as well as auto-indentation. You can use F12 and Shift+F12 for Go-to-definition and Find-All-References, respectively, as well hovering for quickinfo.
+
+## Supported Tags
+
+Any tag may be used for documentation, but Visual Studio treats some tags specially to improve the editor's ability to catch errors and help with refactoring.
+The following tags are handled specially. Each tag is followed by a description and example.
+
+* `@param` -- describes a parameter to a callable construct (eg: a function). Type annotations and descriptions are optional.
+
+```ts
+/**
+ * Repeats a string a certain number of times.
+ * Also: the future is bright and consists of native string repetition:
+ * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/repeat
+ *
+ * @param {string} string  String to repeat
+ * @param {number} times   Number of times to repeat string. Integer.
+ * @see http://jsperf.com/string-repeater/2
+ */
+function repeatString(string, times) {
+  if (times === 1) {
+    return string;
+  }
+  if (times < 0) { throw new Error(); }
+  var repeated = '';
+  while (times) {
+    if (times & 1) {
+      repeated += string;
+    }
+    if (times >>= 1) {
+      string += string;
+    }
+  }
+  return repeated;
+}
+```
+
+* `@return` -- describes the return value of a callable construct (eg: a function). Type annotations and descriptions are optional.
+
+```ts
+/**
+ * Determine if the string ends with the specified substring.
+ *
+ * @param {string} haystack String to search in
+ * @param {string} needle   String to search for
+ * @return {boolean}
+ */
+function endsWith(haystack, needle) {
+  return haystack.slice(-needle.length) === needle;
+}
+```
+
+* `@class` -- tells Visual Studio to treat the next construct as a constructor. (Only available in JavaScript files)
+
+```ts
+var Cell = /** @class */ (function () {
+    function Cell() {
+    }
+    return Cell;
+}());
+```
+
+* `@augments` and `@extends` -- in JsDoc for class-like declarations, specifies which class the upcoming construct extends. Note that `@augments` and `@extends` are synonyms.
+
+```ts
+// @Filename: declarations.d.ts
+ declare class Thing<T> {
+     mine: T;
+ }
+
+// @Filename: dummy.js
+ /**
+  * @augments {Thing<string>}
+  */
+ class MyStringThing extends Thing {
+     constructor() {
+         // x is a string!
+         var x = this.mine;
+     }
+ }
+```
+
+* `@template` -- used to specify type parameters for the following construct (c.f. polymorphism in JavaScript)
```

Comment:
JS programmers think polymorphism is implementation inheritance polymorphism. You should probably say "generic programming".

The example might be easier to understand if you also specify the return type.

Reply:
changed.